Documentation about fixed point types. (#11701)
ethereum/solidity: docs/types/value-types.rst

-------------------

.. warning::
    Fixed point numbers are not fully supported by Solidity yet.

``fixed`` / ``ufixed``: Signed and unsigned fixed point number of various sizes. Keywords ``ufixedMxN`` and ``fixedMxN``, where ``M`` represents the total number of bits used by

> the type and ``N`` represents how many fractional decimal digits are available. ``M`` must be divisible by 8 and goes from 8 to 256 bits. ``N`` must be between 0 and 80, inclusive.

Reviewer: ```suggestion
the type and ``N`` represents how many fractional decimal digits are available. The number ``M`` must be divisible by 8 and goes from 8 to 256 bits and ``N`` must be between 0 and 80, inclusive.
```

Reviewer: I wonder if we need to include `0` for `N`.

Author: It might be useful in the future, because it is clear that the result of `fixed128x2 * fixed64x0` is a fixed point number, while it is not clear for `fixed128x2 * int64`.

But we can also remove it now and re-introduce it when we think it is useful.

``ufixed`` and ``fixed`` are aliases for ``ufixed128x18`` and ``fixed128x18``, respectively.

Fixed point types are used to represent fractional numbers with a fixed precision. The number ``1.5`` is stored in a ``fixed16x2`` (16 bit number with
two fractional digits) as ``1.50``. In the EVM, it is stored exactly the same way as the number ``150`` would be stored in an ``int16``.

> The largest and the smallest number representable in a ``fixed16x2`` are ``327.67`` and ``-327.69``, respectively.

Reviewer: ```suggestion
The largest and the smallest number representable in a ``fixed16x2`` are ``327.67`` and ``-327.68``, respectively.
```

For a fixed point type ``X``, you can use ``type(X).min`` and ``type(X).max`` to

> access the minimum and maximum value representable by the type, e.g. ``type(fixed16x2).max`` is ``327.67``.

Reviewer: Do we keep a list of unimplemented features for fixed points?

Reviewer: I opened this: https://github.com/ethereum/solidity/issues/11719

Type conversions:

Implicit conversions from integer and fixed point types to ``fixedMxN`` types can be performed,
as long as all values can still be represented ("widening" conversion).

Implicit conversion from rational number expressions (literals) to fixed point types are only possible if there
is no precision loss and the number is within the value range. This means that 0.01 can be implicitly
converted to ``fixed64x2`` but not to ``fixed64x1``.

Explicit conversions from rational number expressions (literals) are always possible as long as the
value is within the range of the fixed point type.

.. note::

>     Expressions like ``1/3``, which do not fit finite precision, always have to be explicitly converted to

Reviewer: ```suggestion
    Expressions like ``1/3``, which do not fit any finite precision, always have to be explicitly converted to
```

    a fixed point type to specify the desired precision. Use ``fixed64x4(1/3)`` for example to have
    four decimal digits of precision.

Explicit conversions between fixed point types can be used to reduce precision (resulting in rounding
towards zero) or value range (resulting in removing higher order bits), but only
one of value range, signedness and precision can be changed at a time.

Explicit conversions to integer types truncate the fractional part (towards zero) and can only be performed if the integer
type can contain the resulting value after truncation.

Explicit conversions to and from ``bytesNN`` types of matching width can be performed and do not change the internal representation.

Operators:

* Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
* Arithmetic operators: ``+``, ``-``, unary ``-``, ``*``, ``/``, ``%`` (modulo)

.. note::
    Arithmetic operators are not yet supported.

.. note::
    Many programming language use floating point numbers (``float`` and ``double``, more precisely IEEE 754 numbers)
    to handle fractional values. Floating point numbers use a different mechanism that makes it possible to represent

>     numbers close to zero with higher precision. Since the exact precision is harder to predict, though, they are

Reviewer: > Floating point numbers use a different mechanism that makes it possible to represent
    numbers close to zero with higher precision

I don't really agree with this.

Author: Yes, it depends on whether you mean absolute or relative precision. I'll try to find a better wording.

Author: I updated the paragraph.

    cannot be assigned to or from.

``fixed`` / ``ufixed``: Signed and unsigned fixed point number of various sizes. Keywords ``ufixedMxN`` and ``fixedMxN``, where ``M`` represents the number of bits taken by
the type and ``N`` represents how many fractional decimal digits are available. ``M`` must be divisible by 8 and goes from 8 to 256 bits. ``N`` must be between 0 and 80, inclusive.
``ufixed`` and ``fixed`` are aliases for ``ufixed128x18`` and ``fixed128x18``, respectively.

Fixed point types are used to represent fractional numbers with a fixed precision. The number ``1.5`` is stored in a ``fixed8x2`` (8 bit number with

> two fractional digits) as ``1.50``. In the EVM, it is stored in exactly the same way as the number ``150`` would be stored in an ``int8``.

Reviewer: ```suggestion
two fractional digits) as ``1.50``. In the EVM, it is stored exactly the same way as the number ``150`` would be stored in an ``int8``.
```

> The largest and smallest numbers representable in a ``fixed8x2`` are ``1.27`` and ``-1.28``, respectively.

Reviewer: ```suggestion
The largest and the smallest number representable in a ``fixed8x2`` are ``1.27`` and ``-1.28``, respectively.
```

Type conversions:

Implicit conversions from integer and fixed point types to ``fixedMxN`` types can be performed,
as long as all values can still be represented ("widening" conversion).

> Explicit conversions to integer types truncate the fractional part (towards zero) and can only be performed if the integer

Reviewer 2: How do you imagine this conversion to happen? Do we want this already or should we just let people manually truncate if they need to? (As they mostly do with scaling between various precisions.)

Author: We can leave it out, but if we only allow conversion to bytes type, it might be a bit cumbersome to use. How do you imagine the manual truncation to work?

> type can contain all resulting values after truncation.

Reviewer: ```suggestion
type can contain the resulting value after truncation.
```

> Explicit conversions to and from ``bytesNN`` types of matching width can be performed and do not change the internal representation.

Reviewer: I think we shouldn't do this, at least until users request for this.

Author: Why not? This is the main mechanism to actually do something with fixed point types if we do not allow operators.

Operators:

* Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
* Arithmetic operators: ``+``, ``-``, unary ``-``, ``*``, ``/``, ``%`` (modulo)

.. note::

>     The main difference between floating point (``float`` and ``double`` in many languages, more precisely IEEE 754 numbers) and fixed point numbers is

Reviewer: I would probably add that the main difference between floating point and fixed point is in rounding errors. Something like:

All operations in floating point numbers have associated rounding errors, whereas in fixed point numbers, addition, multiplication (?) and subtraction doesn't have any rounding errors. Rounding errors can be detrimental for smart contracts, and fixed point numbers minimize them.

Not completely sure about errors in multiplication, since it may depend on implementation.

Author: Yeah, maybe we should also just drop this paragraph and only explain how fixed point numbers work and how the precision works.

Reviewer 3: I would not drop it completely. I think it should at least say that these are not floating-point numbers and they work differently. From my experience, many people do not know the difference, even experienced devs who were not taught this formally and just never had a need to dig deeper into this topic on their own. Just instead of explaining what the difference is we could link to some external article.

    that the number of bits used for the integer and the fractional part (the part after the decimal dot) is flexible in the former, while it is static

>     in the latter. Generally, in floating point almost the entire space is used to represent the number, while only a small number of bits define
>     where the decimal point is.

Reviewer: ```suggestion
    in the latter. 
```
I don't think it's fair to compare two different representation systems like that. To be fair, double can do 15-16 decimal digits, when compared to 18 in the usual fixed point.

    cannot be assigned to or from.

``fixed`` / ``ufixed``: Signed and unsigned fixed point number of various sizes. Keywords ``ufixedMxN`` and ``fixedMxN``, where ``M`` represents the number of bits taken by
the type and ``N`` represents how many fractional decimal digits are available. ``M`` must be divisible by 8 and goes from 8 to 256 bits. ``N`` must be between 0 and 80, inclusive.
``ufixed`` and ``fixed`` are aliases for ``ufixed128x18`` and ``fixed128x18``, respectively.

> Fixed point types are used to represent fractional numbers with a fixed precision. The number ``1.5`` is stored in a ``fixed8x2`` (8 bit number with
> two fractional digits) as ``1.50``. In the EVM, it is stored exactly the same way as the number ``150`` would be stored in an ``int8``.
> The largest and the smallest number representable in a ``fixed8x2`` are ``1.27`` and ``-1.28``, respectively.

Reviewer 3: If `fixed8x2` can represent numbers between `-1.28` and `1.27` then `1.5` is out of range. The example should use a smaller number:
```suggestion
Fixed point types are used to represent fractional numbers with a fixed precision. The number ``1.23`` is stored in a ``fixed8x2`` (8 bit number with
two fractional digits) as ``1.23``. In the EVM, it is stored exactly the same way as the number ``123`` would be stored in an ``int8``.
The largest and the smallest number representable in a ``fixed8x2`` are ``1.27`` and ``-1.28``, respectively.
```

Author: haha good catch! :)

Author: I would still try to have a trailing zero

Reviewer: Haha, nice!

Reviewer 3: Good point. 1.20 or something might be better then.

> The largest and the smallest number representable in a ``fixed8x2`` are ``1.27`` and ``-1.28``, respectively.

Reviewer 3: Oh, so it's possible for `N` to be larger than the maximum number of significant digits? I.e. something like `fixed8x5` is still valid and `123` stored in it would mean `0.00123` and the type would represent numbers between `-0.00128` and `0.00127`? That's nice. Might be worth pointing it out (at least saying that such types are valid).

Author: I think a good condition is that `1` has to be in the value range, but I'm actually not sure what we chose in the end.

> Type conversions:

Reviewer 3: Why not make this and `Operators:` below into proper sections?
```suggestion
Type conversions
^^^^^^^^^^^^^^^^
```

Author: Because it's like that in other sections

Reviewer 3: ok. fine then.
